docs/source/operators/scopes.rst

```diff
@@ -0,0 +1,183 @@
+.. _scoles:
+
+Scopes in Jupyter Server
+========================
+
+A scope has a syntax-based design that reveals which resources it provides access to. Resources are
+objects with a type, associated data, relationships to other resources, and a set of methods that
+operate on them.
+
+``<resource>`` in the RBAC scope design refers to the resource name in the Jupyter Server's API
+endpoints in most cases (maybe prefixed with ``/api/``). For instance, ``<resource>`` equal to
+``users`` corresponds to Jupyter Server's API endpoints beginning with ``/api/users``.
```

Comment:
We should use a resource example like `kernels` or `contents` since those are more server-specific.  Are there plans to add `/api/users` or `/api/groups` endpoints?

Reply:
> We should use a resource example like `kernels` or `contents` since those are more server-specific.

I agree, I will change that.

> Are there plans to add `/api/users` or `/api/groups` endpoints?

Yes, [here](https://github.com/jupyter-server/jupyter_server/pull/540/files#diff-fbbdecb56a3a596acfbcc7ef2d22310194d6af54028894570d1d628c7454f252R91) and [there](https://github.com/jupyter-server/jupyter_server/pull/540/files#diff-fbbdecb56a3a596acfbcc7ef2d22310194d6af54028894570d1d628c7454f252R63).

Reply:
Ok. Looking at the Hub PR I see these are purely application-level entities and not associated with the system-level users/groups, other than perhaps by username.  I assume these endpoints will be similar to Hub's but operate against different model definitions?  (E.g., the user model won't have `server` or `servers` attributes, but will it have `kernels`, `terminals`, `contents manager`, `sessions`?)

Will we be introducing a database for these and can the same DB be used between servers, and/or between Hub and Server?

Reply:
> I assume these endpoints will be similar to Hub's but operate against different model definitions? (E.g., the user model won't have `server` or `servers` attributes, but will it have `kernels`, `terminals`, `contents manager`, `sessions`?)

Yes, Hub and Server's resources are very different, but the notions of users and groups are similar for both. I don't think the user model will have `kernels` attribute and such, as I understand it the user model would just provide an identity (something like a `login`), that would be used to query its permissions. But I don't have much experience and I might be wrong.

> Will we be introducing a database for these and can the same DB be used between servers, and/or between Hub and Server?

Good question, I am actually not sure, maybe we should start without a DB to keep it simple?

Also, I am wondering if we should not experiment in a separate repo first. I think everything could be be implemented in a server extension (like in https://github.com/davidbrochart/jupyterlab-auth), and we could only have #165 merged here, which would not change Jupyter Server dramatically yet allow for RBAC.

Reply:
I agree completely.  Moving this to an extension would still preserve the "authorization provider" pluggability that #165 alludes to and makes a lot of sense.   Since the extension would be using the general framework, scope filtering, etc that Hub does, then it seems like it might be worth looking at a design in which Hub could also leverage _it_.  (Don't really know if Hub has the concept of server extensions, etc., or if that's even a viable comment.)

At any rate, this sounds like a great agenda item for the next [Server meeting](https://github.com/jupyter-server/team-compass#weekly-team-meetings) (next Thursday) if you can make it?

Reply:
Thanks Kevin, let's discuss this further at the next Server meeting.

Reply:
> Will we be introducing a database for these and can the same DB be used between servers, and/or between Hub and Server?

Here's some experimentation I did with this authorization layer (awhile ago)—a Hub service that extends the JupyterHub database with group (and project) information: https://github.com/Zsailer/jupyter-projects